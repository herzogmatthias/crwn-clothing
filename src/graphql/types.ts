import { ICartItem } from "../redux/cart/ICartItem";

export interface IHeaderType {
  cartHidden: boolean;
}

export interface ICacheState {
  cartHidden: boolean;
  cartItems: ICartItem[];
  itemCount: number;
}
