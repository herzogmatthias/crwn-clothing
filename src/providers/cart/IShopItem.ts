export interface IShopItem {
  id: number;
  name: string;
  imageUrl: string;
  price: number;
}
