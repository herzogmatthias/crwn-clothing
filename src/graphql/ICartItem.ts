import { IShopItem } from "./IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
