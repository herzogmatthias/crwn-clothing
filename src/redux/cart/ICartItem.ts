import { IShopItem } from "../shop/IShopItem";

export interface ICartItem extends IShopItem {
  quantity: number;
}
