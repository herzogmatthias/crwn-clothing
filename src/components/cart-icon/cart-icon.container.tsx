import { graphql } from "react-apollo";
import React from "react";
import { flowRight } from "lodash";
import CartIcon from "./cart-icon.component";
import { TOGGLE_CART_HIDDEN } from "../../graphql/mutations";
import { GET_ITEM_COUNT } from "../../graphql/resolvers";
import { ApolloCurrentQueryResult } from "apollo-boost";

interface IWithGraphqlProps
  extends ApolloCurrentQueryResult<{ itemCount: number }> {
  toggleCartHidden(): void;
}

const CartIconContainer = ({ data, toggleCartHidden }: IWithGraphqlProps) => {
  return (
    <CartIcon
      itemCount={data!.itemCount}
      toggleCartHidden={toggleCartHidden}
    ></CartIcon>
  );
};

export default flowRight(
  graphql<{}, { itemCount: number }>(GET_ITEM_COUNT),
  graphql<{}, { toggleCartHidden: () => {} }>(TOGGLE_CART_HIDDEN, {
    name: "toggleCartHidden",
  })
)(CartIconContainer);
