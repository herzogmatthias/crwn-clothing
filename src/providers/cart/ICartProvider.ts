import { ICartItem } from "../../redux/cart/ICartItem";
import { IShopItem } from "../../redux/shop/IShopItem";

export interface ICartProvider {
  hidden: boolean;
  total: number;
  toggleHidden(): void;
  cartItems: ICartItem[];
  addItem(item: IShopItem): void;
  removeItem(item: ICartItem): void;
  clearItemFromCart(item: ICartItem): void;
  cartItemsCount: number;
}
