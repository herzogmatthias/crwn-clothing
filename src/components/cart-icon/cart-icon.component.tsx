import * as React from "react";
import { ConnectedProps, connect } from "react-redux";
import { RootState } from "../../redux/root-reducer";
import { selectCartItemsCount } from "../../redux/cart/cart.selectors";
import { createStructuredSelector } from "reselect";
import {
  CartIconContainer,
  ShoppingIconContainer,
  ItemCountContainer,
} from "./cart-icon.styles";
import CartContext from "../../contexts/cart/cart.context";

interface ISelectorProps {
  itemCount: number;
}

type ICartIconProps = ConnectedProps<typeof connector>;

function CartIcon({ itemCount }: ICartIconProps) {
  const { toggleHidden } = React.useContext(CartContext);
  return (
    <CartIconContainer onClick={toggleHidden}>
      <ShoppingIconContainer></ShoppingIconContainer>
      <ItemCountContainer>{itemCount}</ItemCountContainer>
    </CartIconContainer>
  );
}

const mapStateToProps = createStructuredSelector<RootState, ISelectorProps>({
  itemCount: selectCartItemsCount,
});

const connector = connect(mapStateToProps, {});
export default connect(mapStateToProps, {})(CartIcon);
