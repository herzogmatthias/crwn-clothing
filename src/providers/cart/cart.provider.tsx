import React, { createContext, useEffect } from "react";
import { ICartProvider } from "./ICartProvider";
import { ICartItem } from "../../redux/cart/ICartItem";
import {
  addItemToCart,
  removeItemFromCart,
  filterItemFromCart,
  getCartItemsCount,
  getTotalPrice,
} from "../../redux/cart/cart.utils";
import { IShopItem } from "../../redux/shop/IShopItem";

export const CartContext = createContext<ICartProvider>({
  hidden: true,
  total: 0,
  cartItems: [],
  cartItemsCount: 0,
  toggleHidden: () => {},
  addItem: () => {},
  removeItem: () => {},
  clearItemFromCart: () => {},
});

interface ICartProviderProps {
  children: React.ReactNode;
}

const CartProvider = ({ children }: ICartProviderProps) => {
  const [hidden, setHidden] = React.useState(true);

  const [cartItems, setCartItems] = React.useState<ICartItem[]>([]);
  const [cartItemsCount, setCartItemsCount] = React.useState(0);

  const [total, setTotal] = React.useState(0);

  const toggleHidden = () => {
    setHidden(!hidden);
  };

  const addItem = (item: IShopItem) =>
    setCartItems(addItemToCart(cartItems, item));

  const removeItem = (item: ICartItem) =>
    setCartItems(removeItemFromCart(cartItems, item));
  const clearItemFromCart = (item: ICartItem) =>
    setCartItems(filterItemFromCart(cartItems, item));

  useEffect(() => {
    setCartItemsCount(getCartItemsCount(cartItems));
    setTotal(getTotalPrice(cartItems));
  }, [cartItems]);

  return (
    <CartContext.Provider
      value={{
        total,
        addItem,
        hidden,
        toggleHidden,
        cartItems,
        cartItemsCount,
        removeItem,
        clearItemFromCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export default CartProvider;
