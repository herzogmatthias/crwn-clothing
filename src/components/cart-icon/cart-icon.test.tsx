import React from "react";
import { default as ConnectedCartIcon, CartIcon } from "./cart-icon.component";
import { ReactWrapper, mount, ShallowWrapper, shallow } from "enzyme";
import configureStore, { MockStoreEnhanced } from "redux-mock-store";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { ICartItem } from "../../redux/cart/ICartItem";
import { CartIconContainer, ItemCountContainer } from "./cart-icon.styles";
import { ActionCreatorWithoutPayload } from "@reduxjs/toolkit";
import { TOGGLE_CART_HIDDEN } from "../../redux/cart/cart.types";

describe("CONNECTED CARTICON TESTS", () => {
  let wrapper: ReactWrapper;

  const itemCount = 4;
  let mockStore = configureStore([]);
  const cartItems: ICartItem[] = [
    { id: 1, imageUrl: "asdfasdf", name: "hats", price: 23, quantity: 2 },
    { id: 2, imageUrl: "asdfasdf", name: "hats", price: 23, quantity: 2 },
  ];
  let store: MockStoreEnhanced<unknown, {}>;

  beforeEach(() => {
    const mockState = {
      cart: {
        itemCount: 0,
        cartItems: cartItems,
      },
    };

    store = mockStore(mockState);

    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <ConnectedCartIcon />
        </Provider>
      </BrowserRouter>
    );
  });
  test("render CartIcon Component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("check Props", () => {
    expect(wrapper.find(CartIcon).prop("itemCount")).toEqual(itemCount);
  });
});

describe("NOT CONNECTED CARTICON TESTS", () => {
  let wrapper: ShallowWrapper;
  let mockAction: ActionCreatorWithoutPayload<"TOGGLE_CART_HIDDEN">;
  const dispatch = jest.fn();
  beforeEach(() => {
    mockAction = jest.fn(() => {
      dispatch({ type: TOGGLE_CART_HIDDEN, payload: undefined });
    }) as any;

    const mockProps = {
      toggleCartHidden: mockAction,
      itemCount: 4,
    } as any;
    wrapper = shallow(<CartIcon {...mockProps}></CartIcon>);
  });
  test("dispatch action when button is clicked", () => {
    wrapper.find(CartIconContainer).simulate("click");
    expect(mockAction).toHaveBeenCalled();
  });
  test("render the itemCount as the text", () => {
    const itemCount = parseInt(wrapper.find(ItemCountContainer).text());
    expect(itemCount).toBe(4);
  });
});
