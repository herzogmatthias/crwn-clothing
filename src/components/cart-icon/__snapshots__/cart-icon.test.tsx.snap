// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CONNECTED CARTICON TESTS render CartIcon Component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(CartIcon)>
        <CartIcon
          itemCount={4}
          toggleCartHidden={[Function]}
        >
          <styled.div
            onClick={[Function]}
          >
            <div
              className="sc-AxjAm fUffOI"
              onClick={[Function]}
            >
              <Styled(Component)>
                <ForwardRef(SvgShoppingBag)
                  className="sc-AxirZ iKeIkB"
                >
                  <svg
                    className="sc-AxirZ iKeIkB"
                  >
                    shopping-bag.svg
                  </svg>
                </ForwardRef(SvgShoppingBag)>
              </Styled(Component)>
              <styled.span>
                <span
                  className="sc-AxiKw iNislX"
                >
                  4
                </span>
              </styled.span>
            </div>
          </styled.div>
        </CartIcon>
      </Connect(CartIcon)>
    </Provider>
  </Router>
</BrowserRouter>
`;
